### DISCUSSION TITLE
Traverse subtrees with zipper and add traverse_while

### DISCUSSION THREAD

**Discussion by NickNeck** [doorgan/sourceror#47] [Proposals]
Hi @doorgan 

Thanks for your great work.

I wanted to propose to make it possible to traverse subtrees with `Zipper.traverse`. 

I would also suggest the functions `Zipper.traverse_while/2` and `Zipper.traverse_while/3`. This functions would provide `:skip` to skip a subtree and `:halt` to halt the traversing.

What do you think?

**Comment by doorgan** [comment]
I think it's a great addition. I've already had and seen use cases for traversing sub trees, and I can also see `traverse_while` being useful.

I merged your pull request 💜

**Reply by doorgan** [reply]
Just one use case I've seen "in the wild" is the Surface patcher, which has to get the node and rezip to traverse a subtree:
https://github.com/surface-ui/surface/blob/434e884b8e8b90e0662ca81ffa2e850ac81c07dc/lib/mix/tasks/surface/surface.init/ex_patcher_move.ex#L25-L36
